File: src/main/java/dev/slne/surf/lobby/jar/listener/ParkourListener.java
package dev.slne.surf.lobby.jar.listener;

import dev.slne.surf.lobby.jar.JumpAndRunProvider;
import dev.slne.surf.lobby.jar.PluginInstance;
import java.util.Arrays;
import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerQuitEvent;

public class ParkourListener implements Listener {

  private final JumpAndRunProvider jumpAndRunProvider = PluginInstance.instance().jumpAndRunProvider();

  @EventHandler
  public void onMove(PlayerMoveEvent event) {
    if (!event.hasChangedPosition()) {
      return;
    }

    Block toBlock = event.getTo().getBlock();
    Block block = toBlock.getRelative(BlockFace.DOWN);
    Player player = event.getPlayer();
    Block[] jumps = this.jumpAndRunProvider.getLatestJumps(player);

    if (toBlock.getLocation()
        .equals(jumpAndRunProvider.jumpAndRun().getStart().getBlock().getLocation())) {
      jumpAndRunProvider.start(player);
      return;
    }

    if (jumps == null) {
      return;
    }

    if (jumps.length < 2 || jumps[0] == null || jumps[1] == null) {
      return;
    }

    Location playerLocation = player.getLocation();
    if (playerLocation.getY() < jumps[0].getLocation().getY() && playerLocation.getY() < jumps[1].getLocation().getY()) {
      jumpAndRunProvider.remove(player);
      return;
    }

    if (jumps[1] == null) {
      return;
    }

    if (block.equals(jumps[1])) {

      jumps[1].setType(jumpAndRunProvider.blocks().get(player));

      this.jumpAndRunProvider.addPoint(player);
      this.jumpAndRunProvider.checkHighScore(player);
      this.jumpAndRunProvider.generate(player);
    }
  }

  @EventHandler
  public void onInteract(PlayerInteractEvent event) {
    Player player = event.getPlayer();
    Block block = event.getClickedBlock();

    if(block == null) {

    }

    Block[] jumps = PluginInstance.instance().jumpAndRunProvider().getLatestJumps(player);

    for (Block jump : jumps) {
      if(jump.getLocation().equals(block.getLocation())) {
        player.sendBlockChange(jump.getLocation(), jump.getBlockData());
      }
    }
  }

  @EventHandler
  public void onQuit(PlayerQuitEvent event) {
    Player player = event.getPlayer();

    jumpAndRunProvider.onQuit(player);
  }
}

File: src/main/java/dev/slne/surf/lobby/jar/listener/ParkourListener.kt
package dev.slne.surf.lobby.jar.listener

import dev.slne.surf.lobby.jar.JumpAndRunService
import dev.slne.surf.lobby.jar.PluginInstance
import org.bukkit.Location
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerMoveEvent
import org.bukkit.event.player.PlayerQuitEvent

class ParkourListener : Listener {
    @EventHandler
    fun onMove(event: PlayerMoveEvent) {
        if (!event.hasChangedPosition()) {
            return
        }

        val toBlock = event.to.block
        val block = toBlock.getRelative(BlockFace.DOWN)
        val player = event.player
        val jumps = JumpAndRunService.getLatestJumps(player)
        val startLocation: Location = JumpAndRunService.jumpAndRun.start ?: return

        if (toBlock.location == startLocation.block.location) {
            JumpAndRunService.start(player)
            return
        }

        if (jumps.isEmpty() || jumps.size < 2) {
            return
        }

        if (jumps[0] == null || jumps[1] == null) {
            return
        }

        val jump1: Block = jumps[0] ?: return
        val jump2: Block = jumps[1] ?: return



        val playerLocation = player.location
        if (playerLocation.y < jump1.location.y && playerLocation.y < jump2.location.y) {
            JumpAndRunService.remove(player)
            return
        }

        if (block == jumps[1]) {
            val material = JumpAndRunService.blocks[player] ?: return

            jump2.type = material

            JumpAndRunService.addPoint(player)
            JumpAndRunService.checkHighScore(player)
            JumpAndRunService.generate(player)
        }
    }

    @EventHandler
    fun onInteract(event: PlayerInteractEvent) {
        val player = event.player
        val location = event.interactionPoint ?: return

        val jumps: Array<Block?> = JumpAndRunService.getLatestJumps(player)

        for (jump in jumps) {
            if (jump == null) {
                continue
            }

            if (jump.location == location) {
                player.sendBlockChange(jump.location, jump.blockData)
            }
        }
    }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        val player = event.player

        JumpAndRunService.onQuit(player)
    }
}
